### ISSUE TITLE
`set` type

### ISSUE THREAD

**Issue by anton-trunov** [tact-lang/tact#580]
It should support the following operations (it's a rough draft, please comment):

- `emptySet()`: create an empty set;
- `s.exist(e)`: returns `true` if the element `e` is in the set and `false` otherwise;
- `s.add(e)`: modify the set `s` by adding a new element `e` to it (we can discuss what this function should return, for instance it can be a boolean signaling if `e` was already in the set);
- `s.replace(e1, e2)`: remove `e1` and add `e2` -- let's discuss the precise semantics or if this is even needed;
- `s1.union(s2)`: `s1` gets extended with the elements of `s2` (we might want to also introduce a non-modifying version of this: `fun union(s1: set<T>, s2: set<T>): set<T>` where `T` is some "setable" type)
- `s1.intersect(s2)`: `s1` will keep only the elements common between `s1` and `s2` (the non-modifying version can be also added)
- ... (set difference, symmetric difference)
- `foreach` statement support for sets
- `toCell()`, etc.

**Comment by novusnota** [comment]
Would the set be implemented as a `map<Int, Bool>` under the hood?

**Comment by Gusarich** [comment]
This could be useful sometimes for sure!

Some thoughts:
 * there should also be 'remove` method to remove some value from the set (it could also return a boolean indicating whether anything was removed or not)
 * I'm not sure if `replace` method is really needed
 * what about adding possibility to get the set length? similarly to #471

**Comment by Gusarich** [comment]
> Would the set be implemented as a `map<Int, Bool>` under the hood?

not `Bool` but just empty values

**Comment by novusnota** [comment]
> there should also be `remove` method to remove some value from the set (it could also return a boolean indicating whether anything was removed or not)

Yup, and it can be called `.del()` just to have the similar API to maps & be 3-letter word counterpart of `.add`/`.set`

> I'm not sure if replace method is really needed

Yeah, having an overriding `.add()` seems simpler

Also agree about adding the length operation :)

**Comment by anton-trunov** [comment]
> what about adding possibility to get the set length?

yeah, that would be useful for sure (let's also see if we should have sets with lengths as the default or as a separate type)

> Would the set be implemented as a map<Int, Bool> under the hood?

hopefully not :) I always found it confusing when people did that (sets are maps into a (the) unit type a.k.a. `void`, but intrinsically let's see if we can do better)

**Comment by Gusarich** [comment]
> hopefully not :) I always found it confusing when people did that (sets are maps into a (the) unit type a.k.a. `void`, but intrinsically let's see if we can do better)

I don't see any problem with this approach. If we try to implement it ourselves without using TVM dictionaries, it will eventually consume more gas.

**Comment by Gusarich** [comment]
And another suggestion for a method:
 * `s1.difference(s2)` - returns a set that has elements from s1 that are not present in s2

**Comment by Gusarich** [comment]
Implementing set as a tvm dictionary `X -> void` limits possible types for `X`. For example, we won't be able to store multi-cell structures in such sets.

I suggest to implement them either as `int -> X` where keys would increment from 0 under the hood, or as `int -> X` where keys would be calculated as hashes of value cells.

**Comment by anton-trunov** [comment]
> int -> X where keys would increment from 0 under the hood

Hmm, this first option does not seem to be a working solution. A set is something we can ask questions like "does this element contain in you"? So, the user has the element not its index (if I understood the proposal correctly). The hashing scheme should work in this case.

**Comment by Gusarich** [comment]
> > int -> X where keys would increment from 0 under the hood
> 
> Hmm, this first option does not seem to be a working solution. A set is something we can ask questions like "does this element contain in you"? So, the user has the element not its index (if I understood the proposal correctly). The hashing scheme should work in this case.

you're right, I suggested a wrong option. The thing with hashing will work but it'll consume a lot of gas (500+ for each hash computation), so it's not a good thing either..
